File: sdk/go/pkg/a2areg/models_test.go
package a2areg

import (
	"encoding/json"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestAgent_FromJSON(t *testing.T) {
	data := []byte(`{
		"id": "agent-1",
		"name": "Test Agent",
		"description": "A test agent",
		"version": "1.0.0",
		"provider": "test-provider",
		"tags": ["ai", "assistant"],
		"is_public": true,
		"is_active": true
	}`)

	var agent Agent
	err := agent.FromJSON(data)
	require.NoError(t, err)
	assert.Equal(t, "agent-1", *agent.ID)
	assert.Equal(t, "Test Agent", agent.Name)
	assert.Equal(t, "A test agent", agent.Description)
	assert.Equal(t, "1.0.0", agent.Version)
	assert.Equal(t, "test-provider", agent.Provider)
	assert.Equal(t, []string{"ai", "assistant"}, agent.Tags)
	assert.True(t, agent.IsPublic)
	assert.True(t, agent.IsActive)
}

func TestAgent_ToJSON(t *testing.T) {
	id := "agent-1"
	agent := &Agent{
		ID:          &id,
		Name:        "Test Agent",
		Description: "A test agent",
		Version:     "1.0.0",
		Provider:    "test-provider",
		Tags:        []string{"ai", "assistant"},
		IsPublic:    true,
		IsActive:    true,
	}

	data, err := agent.ToJSON()
	require.NoError(t, err)

	var result Agent
	err = json.Unmarshal(data, &result)
	require.NoError(t, err)
	assert.Equal(t, *agent.ID, *result.ID)
	assert.Equal(t, agent.Name, result.Name)
	assert.Equal(t, agent.Description, result.Description)
}

func TestAgent_WithCapabilities(t *testing.T) {
	streaming := true
	pushNotifications := false
	agent := &Agent{
		Name:        "Test Agent",
		Description: "A test agent",
		Version:     "1.0.0",
		Provider:    "test-provider",
		Capabilities: &AgentCapabilities{
			Streaming:         &streaming,
			PushNotifications: &pushNotifications,
		},
	}

	assert.NotNil(t, agent.Capabilities)
	assert.True(t, *agent.Capabilities.Streaming)
	assert.False(t, *agent.Capabilities.PushNotifications)
}

func TestAgent_WithAuthSchemes(t *testing.T) {
	name := "X-API-Key"
	agent := &Agent{
		Name:        "Test Agent",
		Description: "A test agent",
		Version:     "1.0.0",
		Provider:    "test-provider",
		AuthSchemes: []SecurityScheme{
			{
				Type: "apiKey",
				Name: &name,
			},
		},
	}

	assert.Len(t, agent.AuthSchemes, 1)
	assert.Equal(t, "apiKey", agent.AuthSchemes[0].Type)
	assert.Equal(t, "X-API-Key", *agent.AuthSchemes[0].Name)
}

func TestAgentCardSpec_FromJSON(t *testing.T) {
	data := []byte(`{
		"name": "Test Agent Card",
		"description": "Card description",
		"url": "https://test.com",
		"version": "1.0.0",
		"capabilities": {
			"streaming": false
		},
		"securitySchemes": [{
			"type": "apiKey",
			"location": "header",
			"name": "X-API-Key"
		}],
		"skills": [{
			"id": "skill-1",
			"name": "Main Skill",
			"description": "Primary skill",
			"tags": []
		}],
		"interface": {
			"preferredTransport": "jsonrpc",
			"defaultInputModes": ["text/plain"],
			"defaultOutputModes": ["text/plain"]
		}
	}`)

	var card AgentCardSpec
	err := card.FromJSON(data)
	require.NoError(t, err)
	assert.Equal(t, "Test Agent Card", card.Name)
	assert.Equal(t, "Card description", card.Description)
	assert.Len(t, card.SecuritySchemes, 1)
	assert.Len(t, card.Skills, 1)
}

func TestAgentCardSpec_ToJSON(t *testing.T) {
	card := &AgentCardSpec{
		Name:        "Test Agent Card",
		Description: "Card description",
		URL:         "https://test.com",
		Version:     "1.0.0",
		Capabilities: AgentCapabilities{},
		SecuritySchemes: []SecurityScheme{
			{Type: "apiKey"},
		},
		Skills: []AgentSkill{
			{
				ID:          "skill-1",
				Name:        "Main Skill",
				Description: "Primary skill",
				Tags:        []string{},
			},
		},
		Interface: AgentInterface{
			PreferredTransport: "jsonrpc",
			DefaultInputModes:  []string{"text/plain"},
			DefaultOutputModes: []string{"text/plain"},
		},
	}

	data, err := card.ToJSON()
	require.NoError(t, err)

	var result AgentCardSpec
	err = json.Unmarshal(data, &result)
	require.NoError(t, err)
	assert.Equal(t, card.Name, result.Name)
	assert.Equal(t, card.Description, result.Description)
}

func TestAgent_WithTimestamps(t *testing.T) {
	now := time.Now()
	agent := &Agent{
		Name:        "Test Agent",
		Description: "A test agent",
		Version:     "1.0.0",
		Provider:    "test-provider",
		CreatedAt:   &now,
		UpdatedAt:   &now,
	}

	assert.NotNil(t, agent.CreatedAt)
	assert.NotNil(t, agent.UpdatedAt)
	assert.Equal(t, now, *agent.CreatedAt)
	assert.Equal(t, now, *agent.UpdatedAt)
}

File: sdk/typescript/tests/models.test.ts
/**
 * Tests for models and helper functions
 */

import {
  agentFromDict,
  capabilitiesFromDict,
  securitySchemeFromDict,
  agentSkillFromDict,
  agentCardSpecFromDict,
  agentProviderFromDict,
  teeDetailsFromDict,
  agentCardSignatureFromDict,
} from '../src/models';

describe('Models', () => {
  describe('agentFromDict', () => {
    it('should create agent from dictionary', () => {
      const data = {
        id: 'agent-1',
        name: 'Test Agent',
        description: 'A test agent',
        version: '1.0.0',
        provider: 'test-provider',
        tags: ['ai', 'assistant'],
        is_public: true,
        is_active: true,
      };

      const agent = agentFromDict(data);
      expect(agent.id).toBe('agent-1');
      expect(agent.name).toBe('Test Agent');
      expect(agent.description).toBe('A test agent');
      expect(agent.version).toBe('1.0.0');
      expect(agent.provider).toBe('test-provider');
      expect(agent.tags).toEqual(['ai', 'assistant']);
      expect(agent.is_public).toBe(true);
    });

    it('should handle missing optional fields', () => {
      const data = {
        name: 'Test Agent',
        description: 'A test agent',
        version: '1.0.0',
        provider: 'test-provider',
      };

      const agent = agentFromDict(data);
      expect(agent.id).toBeUndefined();
      expect(agent.tags).toEqual([]);
      expect(agent.is_public).toBe(true);
    });

    it('should handle agentId as id', () => {
      const data = {
        agentId: 'agent-123',
        name: 'Test Agent',
        description: 'A test agent',
        version: '1.0.0',
        provider: 'test-provider',
      };

      const agent = agentFromDict(data);
      expect(agent.id).toBe('agent-123');
    });

    it('should handle capabilities', () => {
      const data = {
        name: 'Test Agent',
        description: 'A test agent',
        version: '1.0.0',
        provider: 'test-provider',
        capabilities: {
          streaming: true,
          pushNotifications: false,
        },
      };

      const agent = agentFromDict(data);
      expect(agent.capabilities).toBeDefined();
      expect(agent.capabilities?.streaming).toBe(true);
    });

    it('should handle auth_schemes', () => {
      const data = {
        name: 'Test Agent',
        description: 'A test agent',
        version: '1.0.0',
        provider: 'test-provider',
        auth_schemes: [
          {
            type: 'apiKey',
            location: 'header',
            name: 'X-API-Key',
          },
        ],
      };

      const agent = agentFromDict(data);
      expect(agent.auth_schemes).toHaveLength(1);
      expect(agent.auth_schemes?.[0].type).toBe('apiKey');
    });

    it('should handle skills', () => {
      const data = {
        name: 'Test Agent',
        description: 'A test agent',
        version: '1.0.0',
        provider: 'test-provider',
        skills: [
          {
            id: 'skill-1',
            name: 'Main Skill',
            description: 'Primary skill',
            tags: ['ai'],
          },
        ],
      };

      const agent = agentFromDict(data);
      expect(agent.skills).toHaveLength(1);
      expect(agent.skills?.[0].id).toBe('skill-1');
    });

    it('should handle agent_card', () => {
      const data = {
        name: 'Test Agent',
        description: 'A test agent',
        version: '1.0.0',
        provider: 'test-provider',
        agent_card: {
          name: 'Test Agent Card',
          description: 'Card description',
          url: 'https://test.com',
          version: '1.0.0',
          capabilities: {},
          securitySchemes: [],
          skills: [],
          interface: {
            preferredTransport: 'jsonrpc',
            defaultInputModes: ['text/plain'],
            defaultOutputModes: ['text/plain'],
          },
        },
      };

      const agent = agentFromDict(data);
      expect(agent.agent_card).toBeDefined();
      expect(agent.agent_card?.name).toBe('Test Agent Card');
    });
  });

  describe('capabilitiesFromDict', () => {
    it('should create capabilities from dictionary', () => {
      const data = {
        streaming: true,
        pushNotifications: false,
        stateTransitionHistory: true,
        supportsAuthenticatedExtendedCard: false,
      };

      const capabilities = capabilitiesFromDict(data);
      expect(capabilities.streaming).toBe(true);
      expect(capabilities.pushNotifications).toBe(false);
      expect(capabilities.stateTransitionHistory).toBe(true);
    });
  });

  describe('securitySchemeFromDict', () => {
    it('should create security scheme from dictionary', () => {
      const data = {
        type: 'oauth2',
        location: 'header',
        name: 'Authorization',
        flow: 'client_credentials',
        tokenUrl: 'https://auth.example.com/token',
        scopes: ['read', 'write'],
      };

      const scheme = securitySchemeFromDict(data);
      expect(scheme.type).toBe('oauth2');
      expect(scheme.location).toBe('header');
      expect(scheme.scopes).toEqual(['read', 'write']);
    });

    it('should handle token_url alias', () => {
      const data = {
        type: 'oauth2',
        token_url: 'https://auth.example.com/token',
      };

      const scheme = securitySchemeFromDict(data);
      expect(scheme.tokenUrl).toBe('https://auth.example.com/token');
    });
  });

  describe('agentSkillFromDict', () => {
    it('should create agent skill from dictionary', () => {
      const data = {
        id: 'skill-1',
        name: 'Main Skill',
        description: 'Primary skill',
        tags: ['ai', 'assistant'],
        examples: ['Example 1'],
        inputModes: ['text/plain'],
        outputModes: ['application/json'],
      };

      const skill = agentSkillFromDict(data);
      expect(skill.id).toBe('skill-1');
      expect(skill.name).toBe('Main Skill');
      expect(skill.tags).toEqual(['ai', 'assistant']);
      expect(skill.examples).toEqual(['Example 1']);
    });

    it('should handle input_modes alias', () => {
      const data = {
        id: 'skill-1',
        name: 'Main Skill',
        description: 'Primary skill',
        tags: [],
        input_modes: ['text/plain'],
        output_modes: ['application/json'],
      };

      const skill = agentSkillFromDict(data);
      expect(skill.inputModes).toEqual(['text/plain']);
      expect(skill.outputModes).toEqual(['application/json']);
    });
  });

  describe('agentCardSpecFromDict', () => {
    it('should create agent card spec from dictionary', () => {
      const data = {
        name: 'Test Agent Card',
        description: 'Card description',
        url: 'https://test.com',
        version: '1.0.0',
        capabilities: {
          streaming: false,
        },
        securitySchemes: [
          {
            type: 'apiKey',
            location: 'header',
            name: 'X-API-Key',
          },
        ],
        skills: [
          {
            id: 'skill-1',
            name: 'Main Skill',
            description: 'Primary skill',
            tags: [],
          },
        ],
        interface: {
          preferredTransport: 'jsonrpc',
          defaultInputModes: ['text/plain'],
          defaultOutputModes: ['text/plain'],
        },
        provider: {
          organization: 'Test Org',
          url: 'https://test.org',
        },
      };

      const card = agentCardSpecFromDict(data);
      expect(card.name).toBe('Test Agent Card');
      expect(card.securitySchemes).toHaveLength(1);
      expect(card.skills).toHaveLength(1);
      expect(card.provider?.organization).toBe('Test Org');
    });
  });

  describe('agentProviderFromDict', () => {
    it('should create agent provider from dictionary', () => {
      const data = {
        organization: 'Test Org',
        url: 'https://test.org',
      };

      const provider = agentProviderFromDict(data);
      expect(provider.organization).toBe('Test Org');
      expect(provider.url).toBe('https://test.org');
    });
  });

  describe('teeDetailsFromDict', () => {
    it('should create TEE details from dictionary', () => {
      const data = {
        enabled: true,
        provider: 'Intel SGX',
        attestation: 'required',
      };

      const teeDetails = teeDetailsFromDict(data);
      expect(teeDetails.enabled).toBe(true);
      expect(teeDetails.provider).toBe('Intel SGX');
      expect(teeDetails.attestation).toBe('required');
    });

    it('should default enabled to false', () => {
      const data = {};
      const teeDetails = teeDetailsFromDict(data);
      expect(teeDetails.enabled).toBe(false);
    });
  });

  describe('agentCardSignatureFromDict', () => {
    it('should create agent card signature from dictionary', () => {
      const data = {
        algorithm: 'RS256',
        signature: 'abc123',
        jwksUrl: 'https://example.com/.well-known/jwks.json',
      };

      const signature = agentCardSignatureFromDict(data);
      expect(signature.algorithm).toBe('RS256');
      expect(signature.signature).toBe('abc123');
      expect(signature.jwksUrl).toBe('https://example.com/.well-known/jwks.json');
    });

    it('should handle jwks_url alias', () => {
      const data = {
        algorithm: 'RS256',
        jwks_url: 'https://example.com/.well-known/jwks.json',
      };

      const signature = agentCardSignatureFromDict(data);
      expect(signature.jwksUrl).toBe('https://example.com/.well-known/jwks.json');
    });
  });
});

